### ISSUE TITLE
[Feature] `spawn()` to treat `Interpreter` like `Actor`

### ISSUE THREAD

**Issue by Ivaka** [davidkpiano/xstate#682]
**Description**
Currently it is not possible to pass an `Interpreter` instance to `spawn(...)` and establish a bidirectional communication channel between interpreters, as it is resolved as `Observable` and treated like one, thus only the parent interpreter would get events from child, but no way to send events to child.

**Comment by Andarist** [comment]
Actors have send method, so if you have a reference to your actor you should be able to pass events to it.

**Comment by Ivaka** [comment]
> Actors have send method, so if you have a reference to your actor you should be able to pass events to it.

Yes, but `Interpreter` is treated as `Observable` when `spawn(interpreterInstance)` is invoked. For `Observable` actors `send() = void 0`.
What you are saying is true, I can directly use my interpreter reference as an `Actor` as it implements the interface, but then I am not able to propagate events from the child interpreter to parent with `sendParent` as there is no relationship between them

**Comment by Andarist** [comment]
> Yes, but Interpreter is treated as Observable when spawn(interpreterInstance) is invoked. For Observable actors send() = void 0.

Oh, damn - you are right! The observable check comes first and it wins over the machine one.

The fix seems really straightforward - would you like to take a stab at it and add appropriate tests? If not I'll get to it later.

**Comment by Ivaka** [comment]
> > Yes, but Interpreter is treated as Observable when spawn(interpreterInstance) is invoked. For Observable actors send() = void 0.
> 
> Oh, damn - you are right! The observable check comes first and it wins over the machine one.
> 
> The fix seems really straightforward - would you like to take a stab at it and add appropriate tests? If not I'll get to it later.

Sure, I'll work on this over the weekend :)

**Comment by allforabit** [comment]
It looks like it's the same logic for invoked services. Maybe it would be worthwhile switching it around for these as well and allowing for an instantiated service?

**Comment by Ivaka** [comment]
> The fix seems really straightforward - would you like to take a stab at it and add appropriate tests? If not I'll get to it later.

Maybe I am missing something here, but it seems that an `Interpreter` wouldn't pass the `isMachine` check so it would still be treated as an `Observable`. 
What I was actually trying to achieve when I ran into this issue was to have two instances of `Interpreter` created in separate places in the code, encapsulating some event triggers, and establish a child-parent relationship between them.
A simplified use case might look like this:
```
const ws = new WebSocket(...);
function async createChildInterpreter() {
    const machine = Machine(...); // I would use sendParent somewhere here
    const interpreter = interpret(machine).start();
    ws.onmessage = (data) => {
        interpreter.send(data.event);
    }
    return interpreter;
}

function createParentInterpreter(child: Interpreter < ... > ) {
    const machine = Machine({
        context: {
            actor: undefined
        },
        states: {
            someState: {
                invoke: {
                    src: createChildInterpreter(),
                    onDone: {
                        actions: [
                            assign({
                                actor: (context, event) => spawn(event.data, {
                                    name: 'child'
                                })
                            }),
                            send({
                                type: 'SOME_EVENT'
                            }, {
                                to: 'child'
                            })
                        ]
                    }
                }
            }
        }
    });
    const interpreter = interpret(machine).start();
    return interpreter;
}

```

**Comment by Andarist** [comment]
Ok, I see the issue now - got confused myself around what we actually spawn.

At the moment we only support spawning machines, not interpreters - the interpreter for a spawned machine is created internally. 

With a given example you are somewhat trying to bypass XState's hierarchy. You actually have 2 external services (that child machine, but also a websocket) which you try to wrap as a single one. I would recommend here to just spawn them separately (if needed) and communicate them through a parent machine.

Maybe I haven't understood the real problem - the code example might be simplified. Could you describe in more detail what you are trying to do?

**Comment by Ivaka** [comment]
You got the idea about right :)
The child machine in the example should basically be a wrapper around some transport, in the example websocket, but not necessarily, it could be rabbitmq for example. The idea is to be able to trigger events from some remote service and the child machine would represent the current state of that service. The parent machine could potentially have a couple of such remote service actors and represent an aggregate of their state.

**Comment by Andarist** [comment]
In this case, you should be able to spawn a child machine which in turn would invoke the transport layer immediately - that way you wouldn't have to use inner interpreter directly at all.

**Comment by Ivaka** [comment]
> In this case, you should be able to spawn a child machine which in turn would invoke the transport layer immediately - that way you wouldn't have to use inner interpreter directly at all.

Thanks for the advice :)
I initially tried this, but it didn't work. My guess was that because websocket evens might occur at any time, when calling `sendParent` from an `WebSocket.onmessage` handler the `Interpreter` might be popped off the [serviceStack](https://github.com/davidkpiano/xstate/blob/ff46e91c96631348d2be116daba009da01b2ad63/src/interpreter.ts#L87) so it doesn't properly propagate event to the parent.

**Comment by Andarist** [comment]
Well - you can't just call `sendParent` inside `onmessage` handler - `sendParent` just returns an action object which is a description for something to be executed by the interpreter. On its own it's pure.

I put together a quick demo, take a look at this - https://codesandbox.io/s/tender-perlman-h9ybn
